// @ant-design/pro-form/es/BaseForm/createField.js
import "antd/es/config-provider/style";
import _ConfigProvider from "antd/es/config-provider";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React, { useContext } from 'react';
import { omitUndefined } from '@ant-design/pro-utils';
import { pickProFormItemProps } from '@ant-design/pro-utils';
import classnames from 'classnames';
import { noteOnce } from "rc-util/es/warning";
import FieldContext from '../FieldContext';
import LightWrapper from './LightWrapper';
import ProFormItem from '../components/FormItem';
var WIDTH_SIZE_ENUM = {
  // 适用于短数字，短文本或者选项
  xs: 104,
  s: 216,
  // 适用于较短字段录入、如姓名、电话、ID 等。
  sm: 216,
  m: 328,
  // 标准宽度，适用于大部分字段长度。
  md: 328,
  l: 440,
  // 适用于较长字段录入，如长网址、标签组、文件路径等。
  lg: 440,
  // 适用于长文本录入，如长链接、描述、备注等，通常搭配自适应多行输入框或定高文本域使用。
  xl: 552
};
/**
 * 这个方法的主要作用的帮助 Field 增加 FormItem 同时也会处理 lightFilter
 *
 * @param Field
 * @param config
 */

function createField(Field, config) {
  var FieldWithContext = function FieldWithContext(props) {
    var _rest$fieldProps, _rest$fieldProps2;

    var size = useContext(_ConfigProvider.SizeContext);

    var _ref = config || {},
        valueType = _ref.valueType,
        customLightMode = _ref.customLightMode,
        lightFilterLabelFormatter = _ref.lightFilterLabelFormatter,
        _ref$valuePropName = _ref.valuePropName,
        valuePropName = _ref$valuePropName === void 0 ? 'value' : _ref$valuePropName,
        defaultProps = _ref.defaultProps,
        ignoreWidth = _ref.ignoreWidth,
        defaultFormItemProps = _objectWithoutProperties(_ref, ["valueType", "customLightMode", "lightFilterLabelFormatter", "valuePropName", "defaultProps", "ignoreWidth"]);

    var _defaultProps$props = _objectSpread(_objectSpread({}, defaultProps), props),
        label = _defaultProps$props.label,
        tooltip = _defaultProps$props.tooltip,
        placeholder = _defaultProps$props.placeholder,
        width = _defaultProps$props.width,
        proFieldProps = _defaultProps$props.proFieldProps,
        bordered = _defaultProps$props.bordered,
        messageVariables = _defaultProps$props.messageVariables,
        ignoreFormItem = _defaultProps$props.ignoreFormItem,
        transform = _defaultProps$props.transform,
        readonly = _defaultProps$props.readonly,
        allowClear = _defaultProps$props.allowClear,
        colSize = _defaultProps$props.colSize,
        propsFormItemProps = _defaultProps$props.formItemProps,
        rest = _objectWithoutProperties(_defaultProps$props, ["label", "tooltip", "placeholder", "width", "proFieldProps", "bordered", "messageVariables", "ignoreFormItem", "transform", "readonly", "allowClear", "colSize", "formItemProps"]);
    /** 从 context 中拿到的值 */


    var _React$useContext = React.useContext(FieldContext),
        fieldProps = _React$useContext.fieldProps,
        formItemProps = _React$useContext.formItemProps; // restFormItemProps is user props pass to Form.Item


    var restFormItemProps = pickProFormItemProps(rest);
    var formNeedProps = {
      value: rest.value,
      onChange: rest.onChange
    };

    var realFieldProps = _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, ignoreFormItem ? formNeedProps : {}), {}, {
      disabled: props.disabled,
      placeholder: placeholder
    }, fieldProps || {}), rest.fieldProps || {}), {}, {
      style: _objectSpread(_objectSpread({}, (_rest$fieldProps = rest.fieldProps) === null || _rest$fieldProps === void 0 ? void 0 : _rest$fieldProps.style), fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.style)
    });

    var otherProps = _objectSpread(_objectSpread(_objectSpread(_objectSpread({
      messageVariables: messageVariables
    }, defaultFormItemProps), formItemProps), restFormItemProps), propsFormItemProps);

    noteOnce( // eslint-disable-next-line @typescript-eslint/dot-notation
    !rest['defaultValue'], '请不要在 Form 中使用 defaultXXX。如果需要默认值请使用 initialValues 和 initialValue。');
    var ignoreWidthValueType = ['switch', 'radioButton', 'radio', 'rate'];
    var field = /*#__PURE__*/React.createElement(Field // ProXxx 上面的 props 透传给 FieldProps，可能包含 Field 自定义的 props，
    // 比如 ProFormSelect 的 request
    , _extends({}, rest, {
      fieldProps: omitUndefined(_objectSpread(_objectSpread({
        allowClear: allowClear
      }, realFieldProps), {}, {
        style: omitUndefined(_objectSpread({
          width: width && !WIDTH_SIZE_ENUM[width] ? width : undefined
        }, realFieldProps === null || realFieldProps === void 0 ? void 0 : realFieldProps.style)),
        className: classnames(realFieldProps === null || realFieldProps === void 0 ? void 0 : realFieldProps.className, _defineProperty({
          'pro-field': width && WIDTH_SIZE_ENUM[width]
        }, "pro-field-".concat(width), width && // 有些 valueType 不需要宽度
        !ignoreWidthValueType.includes(props === null || props === void 0 ? void 0 : props.valueType) && !ignoreWidth && WIDTH_SIZE_ENUM[width]))
      })),
      proFieldProps: omitUndefined(_objectSpread({
        mode: readonly ? 'read' : 'edit',
        params: rest.params,
        proFieldKey: "form-field-".concat(otherProps === null || otherProps === void 0 ? void 0 : otherProps.name)
      }, proFieldProps))
    }));
    return /*#__PURE__*/React.createElement(ProFormItem // 全局的提供一个 tip 功能，可以减少代码量
    // 轻量模式下不通过 FormItem 显示 label
    , _extends({
      label: label && (proFieldProps === null || proFieldProps === void 0 ? void 0 : proFieldProps.light) !== true ? label : undefined,
      tooltip: (proFieldProps === null || proFieldProps === void 0 ? void 0 : proFieldProps.light) !== true && tooltip,
      valuePropName: valuePropName
    }, otherProps, {
      ignoreFormItem: ignoreFormItem,
      transform: transform,
      dataFormat: (_rest$fieldProps2 = rest.fieldProps) === null || _rest$fieldProps2 === void 0 ? void 0 : _rest$fieldProps2.format,
      valueType: valueType || rest.valueType,
      messageVariables: _objectSpread({
        label: label
      }, otherProps === null || otherProps === void 0 ? void 0 : otherProps.messageVariables)
    }), /*#__PURE__*/React.createElement(LightWrapper, _extends({}, realFieldProps, {
      allowClear: allowClear,
      bordered: bordered,
      size: size,
      light: proFieldProps === null || proFieldProps === void 0 ? void 0 : proFieldProps.light,
      customLightMode: customLightMode,
      label: label,
      labelFormatter: lightFilterLabelFormatter,
      valuePropName: valuePropName
    }), field));
  };

  return FieldWithContext;
}

export default createField;
