// @umijs/preset-dumi/lib/api-parser/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function parser() {
  const data = _interopRequireWildcard(require("react-docgen-typescript"));

  parser = function parser() {
    return data;
  };

  return data;
}

var _cache = _interopRequireDefault(require("../utils/cache"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr && (typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]); if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

const cacher = new _cache.default(); // ref: https://github.com/styleguidist/react-docgen-typescript/blob/048980a/src/parser.ts#L1110

const DEFAULT_EXPORTS = ['default', '__function', 'Stateless', 'StyledComponentClass', 'StyledComponent', 'FunctionComponent', 'StatelessComponent', 'ForwardRefExoticComponent'];

var _default = (filePath, componentName) => {
  let definitions = cacher.get(filePath);
  const isDefaultRegExp = new RegExp(`^${componentName}$`, 'i');

  if (!definitions) {
    let defaultDefinition;
    definitions = {};
    parser().withCompilerOptions({
      esModuleInterop: true,
      jsx: 'preserve'
    }, {
      savePropValueAsString: true,
      shouldExtractLiteralValuesFromEnum: true,
      shouldRemoveUndefinedFromOptional: true,
      componentNameResolver: source => {
        // use parsed component name from remark pipeline as default export's displayName
        return DEFAULT_EXPORTS.includes(source.getName()) ? componentName : undefined;
      }
    }).parse(filePath).forEach(item => {
      // convert result to IApiDefinition
      const exportName = isDefaultRegExp.test(item.displayName) ? 'default' : item.displayName;
      const props = Object.entries(item.props).map(([identifier, prop]) => {
        const result = {
          identifier
        };
        const fields = ['identifier', 'description', 'type', 'defaultValue', 'required'];
        const localeDescReg = /(?:^|\n+)@description\s+/;
        fields.forEach(field => {
          switch (field) {
            case 'type':
              result.type = prop.type.raw || prop.type.name;
              break;

            case 'description':
              // the workaround way for support locale description
              // detect locale description content, such as @description.zh-CN xxx
              if (localeDescReg.test(prop.description)) {
                // split by @description symbol
                const groups = prop.description.split(localeDescReg).filter(Boolean);
                groups === null || groups === void 0 ? void 0 : groups.forEach(str => {
                  const _str$match = str.match(/^(\.[\w-]+)?\s*([^]*)$/),
                        _str$match2 = _slicedToArray(_str$match, 3),
                        locale = _str$match2[1],
                        content = _str$match2[2];

                  result[`description${locale || ''}`] = content;
                });
              } else if (prop.description) {
                result.description = prop.description;
              }

              break;

            case 'defaultValue':
              if (prop[field]) {
                result.default = prop[field].value;
              }

              break;

            default:
              if (prop[field]) {
                result[field] = prop[field];
              }

          }
        });
        return result;
      });

      if (exportName === 'default') {
        defaultDefinition = props;
      } else {
        definitions[exportName] = props;
      }
    }); // to make sure default export always in the top

    if (defaultDefinition) {
      definitions = Object.assign({
        default: defaultDefinition
      }, definitions);
    }
  }

  cacher.add(filePath, definitions);
  return definitions;
};

exports.default = _default;
